lib/ConsumerImpl.cc

```diff
@@ -1089,12 +1105,13 @@ void ConsumerImpl::acknowledgeAsync(const MessageId& msgId, ResultCallback callb
     const auto& msgIdToAck = pair.first;
     const bool readyToAck = pair.second;
     if (readyToAck) {
-        ackGroupingTrackerPtr_->addAcknowledge(msgIdToAck);
+        ackGroupingTrackerPtr_->addAcknowledge(msgIdToAck, callback);
+    } else {
+        if (callback) {
```

Comment:
Why do direct callback when readyToAck is false? 

If batchIndexAck = false, and the batch message does not have ack all messages, here also needs to wait(if ackReceiptEnabled == true)?

Reply:
Yes. It has the same behavior with the Java client. See https://github.com/apache/pulsar/blob/7a99e74220bdff7a0dff3dbbf3befb2316f3a6c8/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java#L214

Even if the actual CommandAck request is not sent, the shared bit set of these `BatchMessageIdImpl` instances has already been modified, which could be treated as acknowledged.